app/src/androidTest/java/org/kiwix/kiwixmobile/main/KiwixMainActivityTest.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2019 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.main
+
+import androidx.test.rule.ActivityTestRule
+import org.junit.Rule
+import org.junit.Test
+import org.kiwix.kiwixmobile.utils.StandardActions.openDrawer
+
+class KiwixMainActivityTest {
+  @Rule
+  @JvmField
+  var activityTestRule = ActivityTestRule(KiwixMainActivity::class.java)
+
+  @Test
+  fun testKiwixMainActivity() {
+    kiwixMainRobo {
+      clickReaderOnBottomNav()
```

Comment:
It would be nicer to model these as sub robots eg

```
clickOnReader {
}
clickOnLibrary { 
  clickFileTransferIcon {
 }
}
```

this was the style I was going for in the sadly commented out `ZimManageActivityTest`, unfortunately a relic but perhaps can be salvaged someday into a better test

Reply:
I'll give it a rewrite in future to eliminate its flakiness, not now as you once told that automated tests should be used sparingly. So now I'm going to focus on features, unit tests and implementing OPDS.

Implementing OPDS is going to take a huge rewrite and will require a lot of time but I'll love doing it.

Reply:
the implementation is still not so clean here.

`clickReader` should accept a receiver function on a LibraryRobot as it is a transitory function and the instantiation of a Robot is a guarantee that the thing is displayed so we should be able to get rid of `assertReaderScreenDisplayed()`


i am also not 100% on this ut we could structure these click functions like so

```
fun aClickThatNavigates(func:SomeRobot.()->Unit){
clickOnTheThing()
someRobot(::func)
pressBack()
}
```

In this way our test would look like

```
aScreen{
  clickSubScreen {
    assertSomeDataRendered()
  } 
//test flow naturally return to the outer scope of aScreen
...etc
}
```

Reply:
>  `clickReader` should accept a receiver function on a LibraryRobot

Why, LibraryRobot as receiver on clickReader would definitely look confusing, if we really need receiver then better we should make a sub-robot for reader as well(which has nothing other than a single assertion). And having so many sub-robots is going to create a big mess in KiwixMainRobot file, I have no problem in doing so, I can write sub-robots for each screen/component but I'm saying that the robot file will definitely look confusing enough then although the test file will look so cool and clean.

This is the current status, and it works(with comments enabled). What are your views about it, did I reach close enough to what you asked or following that I wrote is just a waste of time.

|TopLevelDestinationTest|KiwixMainRobot[Part-1]|KiwixMainRobot[Part-2]|
|--|--|--|
|![snip](https://user-images.githubusercontent.com/56556919/96557350-c35f3f80-12d7-11eb-9d15-cddce44497c7.png)|![robo](https://user-images.githubusercontent.com/56556919/96557380-cfe39800-12d7-11eb-92be-3c9719efe15f.png)|![stat2](https://user-images.githubusercontent.com/56556919/96557398-d70aa600-12d7-11eb-99c6-5b583db3d3d3.png)|

Reply:
Maybe this is what you want? I want it to write the exact way depicted in the ss, but it does not let me do so and removed lambda ref. makes it look stupid.
|PotentialFix|Governing Code|
|---|---|
|![roborborobrbor](https://user-images.githubusercontent.com/56556919/96560991-9c573c80-12dc-11eb-9bec-b1b9395cc25f.png)|![downrobor](https://user-images.githubusercontent.com/56556919/96561281-fb1cb600-12dc-11eb-912d-1e979ba03c06.png)|

Reply:
It is much harder to review images. We probably don't want all our robots in the same class, I originally did that because there were 2 tabs inside an activity and they felt logically quite close. Here we would end up with every robot in the app under this and the file we be grossly large.

There are a couple more issues. If a UI element is clickable then we supply a function to allow that element to be clicked in a test. If clicking that element causes navigation then we can pass a lambda that is a receiver function on the screen we are about to navigate to eg in `clickOnGlobeIcon` the receiver is DownloadRobot, that is incorrect as the screen it is navigating to is Language so it should be LanguageRobot here.

We do not need to assert a screen is displayed inside a robot eg
```
clickOnGlobeIcon{
  assertLanguageScreenIsDisplayed()
}
```
by instantiating a robot it asserts it is displayed or throws an error. this is done in the `init` block

If we look at the `this`s here
![image](https://user-images.githubusercontent.com/3358282/96565794-bc770380-12bc-11eb-954e-0259e953add4.png)
there should not be multiple of the same type nested and really in this type of test I don't think we should ever see a robot more than once

Reply:
In the talk you linked, Jake Wharton describes robot as a _User to test the app_, then why we are employing so many users to test a single portion?